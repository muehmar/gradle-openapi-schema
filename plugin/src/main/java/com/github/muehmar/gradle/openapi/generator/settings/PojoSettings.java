package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.task.TaskIdentifier;
import com.github.muehmar.gradle.openapi.warnings.Warning;
import com.github.muehmar.gradle.openapi.warnings.WarningsContext;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.List;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings implements Serializable {
  JsonSupport jsonSupport;
  String packageName;
  String suffix;
  StagedBuilderSettings stagedBuilder;
  String builderMethodPrefix;
  boolean enableValidation;
  boolean nonStrictOneOfValidation;

  ValidationApi validationApi;
  List<ClassTypeMapping> classTypeMappings;
  List<FormatTypeMapping> formatTypeMappings;
  EnumDescriptionSettings enumDescriptionSettings;
  GetterSuffixes getterSuffixes;
  ValidationMethods validationMethods;
  List<String> excludeSchemas;
  PojoNameMappings pojoNameMappings;

  TaskIdentifier taskIdentifier;

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableValidation() {
    return enableValidation;
  }

  public boolean isNonStrictOneOfValidation() {
    return nonStrictOneOfValidation;
  }

  public List<ClassTypeMapping> getClassTypeMappingsList() {
    return classTypeMappings;
  }

  public List<FormatTypeMapping> getFormatTypeMappingsList() {
    return formatTypeMappings;
  }

  public List<String> getExcludeSchemasList() {
    return excludeSchemas;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public ExcludedSchemas getExcludedSchemas() {
    return ExcludedSchemas.fromExcludedPojoNames(
        PList.fromIter(excludeSchemas).map(Name::ofString));
  }

  public boolean isEnableStagedBuilder() {
    return stagedBuilder.isEnabled();
  }

  public boolean isDisableStagedBuilder() {
    return !isEnableStagedBuilder();
  }

  public StagedBuilderSettings getStagedBuilder() {
    return stagedBuilder;
  }

  @FieldBuilder(fieldName = "classTypeMappings")
  public static List<ClassTypeMapping> classTypeMappings(
      PList<ClassTypeMapping> classTypeMappings) {
    return classTypeMappings.toArrayList();
  }

  @FieldBuilder(fieldName = "formatTypeMappings")
  public static List<FormatTypeMapping> formatTypeMappings(
      PList<FormatTypeMapping> formatTypeMappings) {
    return formatTypeMappings.toArrayList();
  }

  public TypeMappings getTypeMappings() {
    return new TypeMappings(PList.fromIter(classTypeMappings), PList.fromIter(formatTypeMappings));
  }

  public PojoNameMapping pojoNameMapping() {
    return PList.fromIter(pojoNameMappings.getConstantNameMappings())
        .map(
            constantNameMapping ->
                PojoNameMapping.replaceConstant(
                    constantNameMapping.getConstant(), constantNameMapping.getReplacement()))
        .foldLeft(PojoNameMapping.noMapping(), PojoNameMapping::andThen);
  }

  public void validate() {
    classTypeMappings.forEach(
        classTypeMapping -> {
          if (not(classTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(classTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });

    formatTypeMappings.forEach(
        formatTypeMapping -> {
          if (not(formatTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(formatTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });
  }
}
