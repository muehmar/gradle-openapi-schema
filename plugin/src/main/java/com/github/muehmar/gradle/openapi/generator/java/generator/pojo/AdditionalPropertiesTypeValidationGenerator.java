package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator.assertTrue;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AdditionalPropertiesTypeValidationGenerator {
  private AdditionalPropertiesTypeValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings>
      additionalPropertiesTypeValidationGenerator() {
    return annotatedCorrectTypeMethod()
        .filter(JavaAdditionalProperties::isAllowed)
        .filter(JavaAdditionalProperties::isNotValueAnyType)
        .contraMap(JavaObjectPojo::getAdditionalProperties);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> annotatedCorrectTypeMethod() {
    return JacksonAnnotationGenerator.<JavaAdditionalProperties>jsonIgnore()
        .append(
            assertTrue(
                props ->
                    String.format(
                        "Not all additional properties are instances of %s",
                        props.getType().getQualifiedClassName().getClassName())))
        .append(correctTypeMethod());
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> correctTypeMethod() {
    return MethodGenBuilder.<JavaAdditionalProperties, PojoSettings>create()
        .modifiers(SettingsFunctions::validationMethodModifiers)
        .noGenericTypes()
        .returnType("boolean")
        .methodName("isAllAdditionalPropertiesHaveCorrectType")
        .noArguments()
        .content(
            constant(
                String.format(
                    "return getAdditionalProperties().size() == %s.size();",
                    additionalPropertiesName())))
        .build();
  }
}
