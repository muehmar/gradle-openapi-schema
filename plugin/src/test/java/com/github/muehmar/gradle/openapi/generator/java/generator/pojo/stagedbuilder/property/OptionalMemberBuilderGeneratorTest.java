package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.stagedbuilder.property;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.stagedbuilder.property.OptionalMemberBuilderGenerator.optionalMemberBuilderGenerator;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.stagedbuilder.StagedBuilderVariant;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

@SnapshotTest
class OptionalMemberBuilderGeneratorTest {
  private Expect expect;

  @ParameterizedTest
  @EnumSource(StagedBuilderVariant.class)
  @SnapshotName("allNecessityAndNullabilityVariants")
  void generate_when_allNecessityAndNullabilityVariants_then_correctOutput(
      StagedBuilderVariant variant) {
    final Generator<JavaObjectPojo, PojoSettings> gen = optionalMemberBuilderGenerator(variant);

    final Writer writer =
        gen.generate(
            JavaPojos.allNecessityAndNullabilityVariants(), defaultTestSettings(), javaWriter());

    expect.scenario(variant.name()).toMatchSnapshot(writerSnapshot(writer));
  }

  @ParameterizedTest
  @EnumSource(StagedBuilderVariant.class)
  @SnapshotName("allNecessityAndNullabilityVariantsFullyTypeMapped")
  void generate_when_allNecessityAndNullabilityVariantsFullyTypeMapped_then_correctOutput(
      StagedBuilderVariant variant) {
    final Generator<JavaObjectPojo, PojoSettings> gen = optionalMemberBuilderGenerator(variant);

    final Writer writer =
        gen.generate(
            JavaPojos.allNecessityAndNullabilityVariantsFullyTypeMapped(),
            defaultTestSettings(),
            javaWriter());

    expect.scenario(variant.name()).toMatchSnapshot(writerSnapshot(writer));
  }
}
