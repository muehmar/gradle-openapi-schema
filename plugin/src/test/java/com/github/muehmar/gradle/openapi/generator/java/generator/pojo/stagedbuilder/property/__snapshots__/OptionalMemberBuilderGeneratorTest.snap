allNecessityAndNullabilityVariantsFullyTypeMapped[FULL]=[
com.custom.CustomList
com.custom.CustomMap
com.custom.CustomString
com.github.muehmar.openapi.util.Tristate
java.util.List
java.util.Map
java.util.Optional

public static final class FullOptPropertyBuilder0 {
  private final Builder builder;

  private FullOptPropertyBuilder0(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalStringVal
   */
  public FullOptPropertyBuilder1 setOptionalStringVal(CustomString optionalStringVal) {
    return new FullOptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

  /**
   * OptionalStringVal
   */
  public FullOptPropertyBuilder1 setOptionalStringVal(Optional<CustomString> optionalStringVal) {
    return new FullOptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

}

public static final class FullOptPropertyBuilder1 {
  private final Builder builder;

  private FullOptPropertyBuilder1(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableStringVal
   */
  public FullOptPropertyBuilder2 setOptionalNullableStringVal(CustomString optionalNullableStringVal) {
    return new FullOptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

  /**
   * OptionalNullableStringVal
   */
  public FullOptPropertyBuilder2 setOptionalNullableStringVal(Tristate<CustomString> optionalNullableStringVal) {
    return new FullOptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

}

public static final class FullOptPropertyBuilder2 {
  private final Builder builder;

  private FullOptPropertyBuilder2(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems(CustomList<CustomString> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems(Optional<CustomList<CustomString>> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems_(CustomList<Optional<CustomString>> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems_(Optional<CustomList<Optional<CustomString>>> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

}

public static final class FullOptPropertyBuilder3 {
  private final Builder builder;

  private FullOptPropertyBuilder3(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems(CustomList<CustomString> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems(Tristate<CustomList<CustomString>> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems_(CustomList<Optional<CustomString>> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems_(Tristate<CustomList<Optional<CustomString>>> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }
}

public static final class FullOptPropertyBuilder4 {
  private final Builder builder;

  private FullOptPropertyBuilder4(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalMap
   */
  public FullOptPropertyBuilder5 setOptionalMap(CustomMap<String, CustomString> optionalMap) {
    return new FullOptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

  /**
   * OptionalMap
   */
  public FullOptPropertyBuilder5 setOptionalMap(Optional<CustomMap<String, CustomString>> optionalMap) {
    return new FullOptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

}

public static final class FullOptPropertyBuilder5 {
  private final Builder builder;

  private FullOptPropertyBuilder5(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableMap
   */
  public FullOptPropertyBuilder6 setOptionalNullableMap(CustomMap<String, CustomString> optionalNullableMap) {
    return new FullOptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

  /**
   * OptionalNullableMap
   */
  public FullOptPropertyBuilder6 setOptionalNullableMap(Tristate<CustomMap<String, CustomString>> optionalNullableMap) {
    return new FullOptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

}
]


allNecessityAndNullabilityVariantsFullyTypeMapped[STANDARD]=[
com.custom.CustomList
com.custom.CustomMap
com.custom.CustomString
com.github.muehmar.openapi.util.Tristate
java.util.List
java.util.Map
java.util.Optional

public static final class OptPropertyBuilder0 {
  private final Builder builder;

  private OptPropertyBuilder0(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalStringVal
   */
  public OptPropertyBuilder1 setOptionalStringVal(CustomString optionalStringVal) {
    return new OptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

  /**
   * OptionalStringVal
   */
  public OptPropertyBuilder1 setOptionalStringVal(Optional<CustomString> optionalStringVal) {
    return new OptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

}

public static final class OptPropertyBuilder1 {
  private final Builder builder;

  private OptPropertyBuilder1(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableStringVal
   */
  public OptPropertyBuilder2 setOptionalNullableStringVal(CustomString optionalNullableStringVal) {
    return new OptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

  /**
   * OptionalNullableStringVal
   */
  public OptPropertyBuilder2 setOptionalNullableStringVal(Tristate<CustomString> optionalNullableStringVal) {
    return new OptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

}

public static final class OptPropertyBuilder2 {
  private final Builder builder;

  private OptPropertyBuilder2(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems(CustomList<CustomString> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems(Optional<CustomList<CustomString>> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems_(CustomList<Optional<CustomString>> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems_(Optional<CustomList<Optional<CustomString>>> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

}

public static final class OptPropertyBuilder3 {
  private final Builder builder;

  private OptPropertyBuilder3(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems(CustomList<CustomString> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems(Tristate<CustomList<CustomString>> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems_(CustomList<Optional<CustomString>> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems_(Tristate<CustomList<Optional<CustomString>>> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }
}

public static final class OptPropertyBuilder4 {
  private final Builder builder;

  private OptPropertyBuilder4(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalMap
   */
  public OptPropertyBuilder5 setOptionalMap(CustomMap<String, CustomString> optionalMap) {
    return new OptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

  /**
   * OptionalMap
   */
  public OptPropertyBuilder5 setOptionalMap(Optional<CustomMap<String, CustomString>> optionalMap) {
    return new OptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

}

public static final class OptPropertyBuilder5 {
  private final Builder builder;

  private OptPropertyBuilder5(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableMap
   */
  public OptPropertyBuilder6 setOptionalNullableMap(CustomMap<String, CustomString> optionalNullableMap) {
    return new OptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

  /**
   * OptionalNullableMap
   */
  public OptPropertyBuilder6 setOptionalNullableMap(Tristate<CustomMap<String, CustomString>> optionalNullableMap) {
    return new OptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

}
]


allNecessityAndNullabilityVariants[FULL]=[
com.github.muehmar.openapi.util.Tristate
java.util.List
java.util.Map
java.util.Optional

public static final class FullOptPropertyBuilder0 {
  private final Builder builder;

  private FullOptPropertyBuilder0(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalStringVal
   */
  public FullOptPropertyBuilder1 setOptionalStringVal(String optionalStringVal) {
    return new FullOptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

  /**
   * OptionalStringVal
   */
  public FullOptPropertyBuilder1 setOptionalStringVal(Optional<String> optionalStringVal) {
    return new FullOptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

}

public static final class FullOptPropertyBuilder1 {
  private final Builder builder;

  private FullOptPropertyBuilder1(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableStringVal
   */
  public FullOptPropertyBuilder2 setOptionalNullableStringVal(String optionalNullableStringVal) {
    return new FullOptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

  /**
   * OptionalNullableStringVal
   */
  public FullOptPropertyBuilder2 setOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
    return new FullOptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

}

public static final class FullOptPropertyBuilder2 {
  private final Builder builder;

  private FullOptPropertyBuilder2(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems(List<String> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems(Optional<List<String>> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems_(List<Optional<String>> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public FullOptPropertyBuilder3 setOptionalListWithNullableItems_(Optional<List<Optional<String>>> optionalListWithNullableItems) {
    return new FullOptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

}

public static final class FullOptPropertyBuilder3 {
  private final Builder builder;

  private FullOptPropertyBuilder3(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems(List<String> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems(Tristate<List<String>> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems_(List<Optional<String>> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public FullOptPropertyBuilder4 setOptionalNullableListWithNullableItems_(Tristate<List<Optional<String>>> optionalNullableListWithNullableItems) {
    return new FullOptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }
}

public static final class FullOptPropertyBuilder4 {
  private final Builder builder;

  private FullOptPropertyBuilder4(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalMap
   */
  public FullOptPropertyBuilder5 setOptionalMap(Map<String, String> optionalMap) {
    return new FullOptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

  /**
   * OptionalMap
   */
  public FullOptPropertyBuilder5 setOptionalMap(Optional<Map<String, String>> optionalMap) {
    return new FullOptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

}

public static final class FullOptPropertyBuilder5 {
  private final Builder builder;

  private FullOptPropertyBuilder5(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableMap
   */
  public FullOptPropertyBuilder6 setOptionalNullableMap(Map<String, String> optionalNullableMap) {
    return new FullOptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

  /**
   * OptionalNullableMap
   */
  public FullOptPropertyBuilder6 setOptionalNullableMap(Tristate<Map<String, String>> optionalNullableMap) {
    return new FullOptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

}
]


allNecessityAndNullabilityVariants[STANDARD]=[
com.github.muehmar.openapi.util.Tristate
java.util.List
java.util.Map
java.util.Optional

public static final class OptPropertyBuilder0 {
  private final Builder builder;

  private OptPropertyBuilder0(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalStringVal
   */
  public OptPropertyBuilder1 setOptionalStringVal(String optionalStringVal) {
    return new OptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

  /**
   * OptionalStringVal
   */
  public OptPropertyBuilder1 setOptionalStringVal(Optional<String> optionalStringVal) {
    return new OptPropertyBuilder1(builder.setOptionalStringVal(optionalStringVal));
  }

}

public static final class OptPropertyBuilder1 {
  private final Builder builder;

  private OptPropertyBuilder1(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableStringVal
   */
  public OptPropertyBuilder2 setOptionalNullableStringVal(String optionalNullableStringVal) {
    return new OptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

  /**
   * OptionalNullableStringVal
   */
  public OptPropertyBuilder2 setOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
    return new OptPropertyBuilder2(builder.setOptionalNullableStringVal(optionalNullableStringVal));
  }

}

public static final class OptPropertyBuilder2 {
  private final Builder builder;

  private OptPropertyBuilder2(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems(List<String> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems(Optional<List<String>> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems_(List<Optional<String>> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

  /**
   * OptionalListWithNullableItems
   */
  public OptPropertyBuilder3 setOptionalListWithNullableItems_(Optional<List<Optional<String>>> optionalListWithNullableItems) {
    return new OptPropertyBuilder3(builder.setOptionalListWithNullableItems_(optionalListWithNullableItems));
  }

}

public static final class OptPropertyBuilder3 {
  private final Builder builder;

  private OptPropertyBuilder3(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems(List<String> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems(Tristate<List<String>> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems_(List<Optional<String>> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }

  /**
   * OptionalNullableListWithNullableItems
   */
  public OptPropertyBuilder4 setOptionalNullableListWithNullableItems_(Tristate<List<Optional<String>>> optionalNullableListWithNullableItems) {
    return new OptPropertyBuilder4(builder.setOptionalNullableListWithNullableItems_(optionalNullableListWithNullableItems));
  }
}

public static final class OptPropertyBuilder4 {
  private final Builder builder;

  private OptPropertyBuilder4(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalMap
   */
  public OptPropertyBuilder5 setOptionalMap(Map<String, String> optionalMap) {
    return new OptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

  /**
   * OptionalMap
   */
  public OptPropertyBuilder5 setOptionalMap(Optional<Map<String, String>> optionalMap) {
    return new OptPropertyBuilder5(builder.setOptionalMap(optionalMap));
  }

}

public static final class OptPropertyBuilder5 {
  private final Builder builder;

  private OptPropertyBuilder5(Builder builder) {
    this.builder = builder;
  }

  /**
   * OptionalNullableMap
   */
  public OptPropertyBuilder6 setOptionalNullableMap(Map<String, String> optionalNullableMap) {
    return new OptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

  /**
   * OptionalNullableMap
   */
  public OptPropertyBuilder6 setOptionalNullableMap(Tristate<Map<String, String>> optionalNullableMap) {
    return new OptPropertyBuilder6(builder.setOptionalNullableMap(optionalNullableMap));
  }

}
]